Implement updated Writable spec

MikaelSmith opened the pull request:
Implement the new Writable interface (defined to mirror the Readable interface). Implement Write on simple file interfaces: local, AWS S3, and Google Cloud Storage.

Resolves #620. Also fixes #656.

MikaelSmith committed:
Make apifs a plugin and include it as the 'local' plugin

This is a developer-centric option that has to be explicitly enabled
with an environment variable.

By default it mounts `/tmp`, can be configured with `local.basepath` to
mount any other directory.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Rework Writable interface to only do full file writes

Implement the new Writable interface (defined to mirror the Readable
interface). Implement Write on a local file interface for testing.

Also adds fuse/file tests.

Supports #620.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Determine size on Open if unknown

This makes editors that rely on size read the entire file. Saves size
from the last open on the fs node so we can use it to serve attribute
requests.

Fixes #656.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

ekinanp commented on api/fs/root.go:
```diff
@@ -0,0 +1,39 @@
+package apifs
```
Why does this need to be a plugin?

MikaelSmith replied:
Just added a comment addressing this and other work that still needs to be done.

ekinanp replied:
Ah, for testing FUSE stuff. Should probably be a developer thing then, i.e. users should not be able to enable it.

MikaelSmith commented:
This is still a work in progress, but I think the approach is about right. Includes a rewrite of fuse/file to broadly mimic https://github.com/bazil/bolt-mount/blob/master/file.go, as suggested in https://github.com/bazil/fuse/issues/225.

Remaining work:
- [ ] Remove or make the `local` plugin disabled by default. It was very useful for testing.
- [ ] Implement `Write` on AWS S3 objects.
- [ ] Add more tests for fuse/file.
- [ ] Revisit TODOs on reading data when file is opened for writing.

ekinanp commented on plugin/types.go:
```diff
@@ -195,12 +195,10 @@ type Readable interface {
-// Writable is an entry that can write data directly to the entry. It mirrors
-// the WriterAt interface with an added context for handling the lifecycle of
-// remote write operations.
+// Writable is an entry that overwrites the prior contents
```
Do you plan on updating the `Writable` docs separately? If not, then this should mention the last bullet point in https://github.com/puppetlabs/wash/issues/620#issuecomment-564192168 since the current comment is not true for GCP topics.

MikaelSmith replied:
I'll make some updates. I will need to write updates for the docs related to external plugins.

ekinanp commented on plugin/gcp/storageObject.go:
```diff
@@ -45,6 +45,16 @@ func (s *storageObject) Read(ctx context.Context, size int64, offset int64) ([]b
+		return err
+	}
+
+	return wr.Close()
```
If `wr.Close()` fails, does that mean the data isn't written (whereby returning a Write error makes sense)? If not, then this should probably be logged to the activity journal or something similar.

MikaelSmith replied:
Yes, it means the data wasn't successfully written.

ekinanp commented on cmd/internal/server/core.go:
```diff
@@ -23,12 +24,19 @@ import (
+		"gcp":        &gcp.Root{},
+		"kubernetes": &kubernetes.Root{},
+	}
+	if os.Getenv("WASH_LOCALFS") != "" {
```
Now that https://github.com/puppetlabs/wash/issues/455 is resolved, I think this should move over to https://github.com/puppetlabs/wash/blob/master/cmd/server.go#L137. That way, you can just say "You can enable localfs by setting WASH_LOCALFS to true" or something w/o having people mess with their `plugins` key.

ekinanp commented on fuse/dir.go:
```diff
@@ -66,7 +67,9 @@ func (d *dir) Lookup(ctx context.Context, req *fuse.LookupRequest, resp *fuse.Lo
 	log.Debugf("FUSE: Found file %v/%v", d, cname)
-	return newFile(d, entry), nil
+	f := newFile(d, entry)
+	f.fillAttr(&resp.Attr)
```
What's the benefit of doing this here?

MikaelSmith replied:
May avoid a separate call to Attr, but it's probably not necessary.

ekinanp reviewed (commented):
Still reviewing, just left a few initial comments.

ekinanp commented on plugin/gcp/pubsubTopic.go:
```diff
@@ -141,14 +141,11 @@ func (t *pubsubTopic) Stream(ctx context.Context) (io.ReadCloser, error) {
 }
 
-func (t *pubsubTopic) Write(ctx context.Context, _ int64, b []byte) (int, error) {
+func (t *pubsubTopic) Write(ctx context.Context, b []byte) error {
```
Similar Q. to https://github.com/puppetlabs/wash/pull/664#discussion_r358519024 regarding result.Get

MikaelSmith replied:
Yes.

ekinanp commented on api/fs/root.go:
```diff
@@ -0,0 +1,55 @@
+
+var _ = plugin.Root(&Root{})
+
+const rootDescription = `
```
Per https://github.com/puppetlabs/wash/pull/664#discussion_r358520344, you could probably capture the config in the LOCALFS environment variable.

MikaelSmith replied:
So some of the discussion in https://github.com/puppetlabs/wash/issues/438 pointed to making apifs just another plugin. That's why I'm implementing it so closely to one.

ekinanp replied:
You could still keep `Init` as-is, you'd just create the config in `server.go` from the env. var.

MikaelSmith replied:
Done.

MikaelSmith committed:
Implement Write for AWS S3 and Google Cloud Storage objects

Also use a context for S3 get object requests.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Implement Writable for external plugins

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
I think this is ready for review again.

MikaelSmith committed:
Update README based on features we've implemented recently

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented on fuse/file.go:
```diff
@@ -3,103 +3,295 @@ package fuse
-			// The entry's content size is unknown so open the file in direct
-			// IO mode. This enables FUSE to still read the entry's content so
-			// that built-in tools like cat and grep still work.
+	if req.Flags.IsReadOnly() || req.Flags.IsReadWrite() {
```
I'm not sure these are the right conditions. We definitely don't want to call `plugin.Size` on an entry that doesn't support Read. We definitely do want the size if opening it ReadOnly, to address how Vim opens it.

If you try to do a write in the middle of a file that doesn't have a Size attribute defined, then you may end up truncating the file (because we won't realize some data is missing that we need to read before writing, see the implementation of `Flush`). So if the file implements ReadAction but is opened WriteOnly, maybe we still want to determine the size. Or maybe we want to error if you try to open a file for writing and haven't explicitly set the Size attribute.

ekinanp reviewed (commented):
Some initial comments, still reviewing

MikaelSmith commented on fuse/file.go:
```diff
@@ -3,103 +3,295 @@ package fuse
-	if closer, ok := fh.r.(io.Closer); ok {
-		return closer.Close()
+	// If this handle had an open writer, write current data.
+	if _, ok := f.writers[req.Handle]; ok {
```
It's not uncommon to call Flush on a handle opened for reading (not sure why). It would also not have an open writer if you opened it for writing but didn't do any calls `write` calls.

MikaelSmith commented on fuse/file.go:
```diff
@@ -3,103 +3,295 @@ package fuse
 
-	if closer, ok := fh.w.(io.Closer); ok {
-		return closer.Close()
+	if len(f.writers) == 0 && releasedWriter {
```
I couldn't think of a way to identify when we'd closed the last open writer. But I could probably delay deleting it from `f.writers` until release as well. We might get some redundant writes if someone decides to call `flush` multiple times, but that's what they asked for.

MikaelSmith committed:
Address feedback and fix issues around tracking file size

Add `sizeValid` to track whether we need to call `plugin.Size` before
using `size`. This is only needed when opening a file WriteOnly to delay
a Read - to ensure we have the entire file content before Write - so we
only do it when necessary.

Handle `io.EOF` when loading the file in case a call to `Setattr`
increased it beyond the original size but didn't write to fill that
space, and pad with null characters when that happens.

Upgrade journal entries to warnings when they result in an error. FUSE
doesn't relay the actual error messages.

Delay releasing data and clearing entry cache until we release all
writers.

Tweak a few tests so we have a balance of ReadWrite tests that rely on
attribute size or calling Read to determine size. Update
TestTruncateAndWrite to ensure we don't call Read when we've called
Setattr to truncate the file first.

Add additional comments explaining how to use BlockReadable and
Writable, and some code comments to clarify how Open works.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented on api/fs/root.go:
```diff
@@ -0,0 +1,52 @@
+	"github.com/puppetlabs/wash/plugin"
+)
+
+// Root of the GCP plugin
```
The rebase was just to fix this. All other changes are in a new commit at the end.

ekinanp reviewed (commented):
Per the [conversation on Slack](https://puppetcommunity.slack.com/archives/CHMCR172S/p1576703313033800), we want a way to distinguish between `file-like` and `non-file-like` entries. 

* `File-like` entries are entries with content, such that Read returns that content and Write updates it. `Non-file-like` entries are entries without content, so that "Read" data could represent one thing and the "Written" data could represent another.

* **For plugin author docs**: `file-like` entries are entries with a `Size` attribute (implicitly entry is also Readable). So an entry has content iff it has a `Size` attribute. Thus, people should not set size for `non-file-like` entries.
    * Reason: `Size` attribute means you can filter those things with `find`. When would filtering on `Size` make sense for Docker container logs, cloud function logs, etc?
        * We could always change these semantics later to a `is-file-like` key in the entry JSON on user demand

* **Implementation**:
    * `File-like` entries are `Readable`, `Writable`, and have a `Size` attribute.
        * Reason: Up to now, we tell people “set size if you know it” for Readable entries. We also do that for things like `metadata.json` entries, which are not `file-like`. These semantics ensure that those entries don't accidentally become `file-like`.


    * For `file-like` entries, `f.data` represents the content. For `non-file-like` entries, `f.data` represents what's about to be written.
    
    * For `file-like` entries, the first `Write` call loads the entire data (because we're going to overwrite it anyways on a `Flush`). `f.data` is then updated appropriately with the written block.
    
    * `Flush` should just write whatever `f.data` is.
    
    * For `Read`, `f.data` should only be returned for `file-like` entries (and if it exists). Otherwise, we should delegate to `plugin.ReadWithAnalytics`. That's because `f.data` represents the content for `file-like` entries while for other entries, it's separate from what could be read.

ekinanp commented:
One example of a non-file like entry that may have Read/Write could be a GCP log, where read could return its most recent 1000 entries while write could write a new entry.

MikaelSmith committed:
Update handling of files without Size attribute

Clarifies difference in how files with and without Size attributes work
by defining them as file-like and non-file-like and describing what that
means. Primarily enables interaction with files where read and write are
not symmetric.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
Updated.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+			}
+			f.data = append(f.data, data...)
+
+			// If still too small then something increased the size beyond the original.
```
When/how could this happen? Should this be a known issue/something we document?

MikaelSmith replied:
Would've been a call to Setattr. Rewriting this comment.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+		}
+
+		// Ensure handle is in list of writers.
+		f.writers[req.Handle] = struct{}{}
```
Might be helpful to mention somewhere that this will likely be (only) called to truncate the file for writes. That helps clarify why we're including the handle in the list of writers.

MikaelSmith replied:
I'll elaborate in the comment. It's not only used to truncate the file (to size 0), you could directly call `truncate` on it and supply any size.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+	attr := plugin.Attributes(f.entry)
+	applyAttr(a, attr, getFileMode(f.entry))
+
+	if f.useLocalState() || !attr.HasSize() {
```
I'd mention that `!attr.HasSize()` should only be true for non-file-like entries

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+	return nil
+}
+
+func (f *file) load(ctx context.Context, start, end int64) ([]byte, error) {
```
This should probably panic if called on a non-file-like entry.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+	}
+}
+
+func getFileMode(entry plugin.Entry) os.FileMode {
```
Nit: `getFileMode` => `defaultMode`. Latter matches the corresponding param. in applyAttr and it seems to be what this is calculating.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+	activity.Record(ctx, "FUSE: Open %v: %+v", f, *req)
+
+	if !f.useLocalState() {
+		// Check for an updated entry in case it has static state.
```
Why do we need to refresh the entry here?

MikaelSmith replied:
Preloaded external plugin content. I'll mention that in the comment.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+}
+
+// If currently writing a file-like object, we should use local state to fulfil many requests.
+func (f *file) useLocalState() bool {
```
I'd also comment that this returning false means "entry is not file-like OR we're not writing to it". Helps clarify some of the `!useLocalState` uses elsewhere in the code.

ekinanp commented on plugin/externalPluginEntry.go:
```diff
@@ -390,6 +391,30 @@ func (e *externalPluginEntry) blockRead(ctx context.Context, size int64, offset
 	return inv.stdout.Bytes(), nil
 }
 
+func (e *externalPluginEntry) Write(ctx context.Context, p []byte) error {
```
We could DRY this up by re-factoring invocation to something like
```
type invocation struct {
  internal.Command // So we get the "SetStdout", "SetStderr", "SetStdin" stuff for free
  // All the other stuff
}

// Add an InvokeAndWait helper to invocation and modify pluginScript's InvokeAndWait to call that.
// May also need to store context in invocation to record to activity?
```

Then in Write, this would read as
```
inv := e.script.NewInvocation(ctx, "write", e).SetStdout(...).SetStderr(...).SetStdin(...)
err := inv.InvokeAndWait()
etc.
```

ekinanp replied:
Maybe we could move over `SetStdout`, `SetStderr` to invocation#InvokeAndWait. Idea is `invocation#InvokeAndWait` would fill invocation#stdout, invocation#stderr.

MikaelSmith replied:
I can work on refactoring things a bit.

ekinanp reviewed (commented):
Reviewing the tests next

ekinanp commented on docs/docs/index.md:
````diff
@@ -260,6 +261,21 @@ wash . ❯ grep "Hello" gcp/Wash/storage/some-wash-stuff/an\ example\ folder/sta
+Note that Wash distinguishes between file-like and non-file-like entries. An entry is file-like if it's readable and writable and defines its size; you can edit it like a file. If it doesn't define a size then it's non-file-like, and trying to open it with a ReadWrite handle will error; reads from it may not return data you previously wrote to it.
+
+#### Examples
+```
````
Actually, would be cool to also include a vim example with a file-like entry (just to cement the fact that it would work with vim).

ekinanp commented on fuse/file_test.go:
```diff
@@ -0,0 +1,653 @@
+	mrw := plugintest.NewMockReadWrite()
+	mrw.Attributes().SetSize(1)
+
+	// expectErrs corresponds to expectations on fields in flags.
```
This took me a bit to figure out what you mean, so I think an example would be useful here.

ekinanp commented on fuse/file_test.go:
```diff
@@ -0,0 +1,653 @@
+	plugin.UnsetTestCache()
+}
+
+func (suite *fileTestSuite) TestOpen_File() {
```
Nit: `File` => `FileLikeEntry` ? Latter's more precise and matches what's in the code.

MikaelSmith replied:
And `TestOpen_NonFile` to `TestOpen_NonFileLikeEntry`? FileLikeEntry is a mouthful, I was using an obvious short-hand.

ekinanp replied:
Doesn't look too bad, but it is a nit so am OK if you want to leave it as-is.

ekinanp replied:
Actually, it does help make the other tests a bit easier to read for me. Like `Read_File` is better understood as `Read_FileLikeEntry`. So would be nice to make that change.

ekinanp replied:
It'd also help to name this test `TestOpen_FileLikeEntry_OpenFlags` since you're testing that Open does the right thing for all combos of openflag and entry. Similar case for the other one.

MikaelSmith replied:
Done.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+	return &file{fuseNode: newFuseNode("f", p, e), writers: make(map[fuse.HandleID]struct{})}
+}
+
+func (f *file) isFileLike() bool {
```
Hmm, maybe `isFileLike` => `isFileLikeEntry` ? It's slightly weird seeing `file#isFileLike`.

MikaelSmith replied:
I don't think we should restrict the definition to only things that are Writable. I think `metadata.json` specifically is file-like, because it has a consistent size.

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
-	if plugin.WriteAction().IsSupportedOn(updatedEntry) {
-		fh.w = updatedEntry.(plugin.Writable)
+	if !f.isFileLike() && req.Flags.IsReadWrite() {
+		// Error ReadWrite on non-file-like entries because it probably won't work well.
```
So with this code, if I have a non-file-like entry that implements both Read/Write, and I do something like `vim <entry>`, I'm going to see what's returned by read b/c vim opens ReadOnly. Then I do some edits and write what I see via `:w`. That's going to be written because vim opens WriteOnly (but what's written may also include the Read data, which can cause some weird issues to happen). I don't think there's anything we can do to prevent that, so might be worth adding a warning for users to be careful when editing non-file-like entries (and preferably that they _shouldn't_ edit non-file-like entries).

MikaelSmith replied:
Editing with vim will work, it just may do something you don't expect. Not sure what to say beyond what's already in the docs.

The comment here is because if you have a ReadWrite handle, you expect what you write to be directly reflected in read operations, and things like `seek` to make sense. They won't, because reads and writes are not symmetric for non-file-like entries.

ekinanp replied:
Oh, the code makes sense. I only commented what I said here b/c that's when I thought of it.

See https://github.com/puppetlabs/wash/pull/664/files#r362674541 re: where to add a note about unexpected things w/ editors.

MikaelSmith commented on api/fs/root.go:
```diff
@@ -0,0 +1,52 @@
+// Schema returns the root's schema
+func (r *Root) Schema() *plugin.EntrySchema {
+	return plugin.
+		NewEntrySchema(r, "local").
```
I think I'm going to skip this one, not an issue for now.

MikaelSmith commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+// file's implementation differs primarily on two axes: whether `Size` is set, and whether we're
+// currently writing.
+// - a *file-like* entry declares a `Size` in its `Attributes`
+//   - reads and writes are symmetric; the kernel page cache will be used, and the file size
```
I disagree slightly because size does not necessary correspond to state. It's possible to grow a file with setattr and then write the content. But it's close enough that I'm fine calling it `content`.

ekinanp replied:
"current local content" => "current local content size". I'd also mention what you just said, so something like "If writing, then this is typically len(data); the atypical case is when the file's truncated via a setattr call"

MikaelSmith replied:
Reworked that section a bit.

MikaelSmith commented on fuse/file.go:
```diff
@@ -3,103 +3,370 @@ package fuse
+//   - `data` stores only the data to be written, and is not initialized from `plugin.Read`
+//   - the file's size will be reported as its readable size; it will not reflect calls to `write`
+//
+// Note that writes only result in calling `plugin.Write` when a file handle is closed. Writing
```
No. This is because FUSE's semantics are confusing. When the system call `close` is called on a file handle, FUSE calls the Flush callback. When the last copy of a file handle is closed, FUSE calls the Release callback. See the description at https://libfuse.github.io/doxygen/structfuse__operations.html#ad4ec9c309072a92dd82ddb20efa4ab14. Would a link to that help?

ekinanp replied:
Yes, that would help.

MikaelSmith replied:
Done.

MikaelSmith committed:
Work on code comments and naming things

Address review comments to try and clarify code.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Extract invocation setup and run into new method

Moves script `InvokeAndWait` behavior to a function on the invocation so
we can setup more complicated invocations before running them. Keeps the
script `InvokeAndWait` as a helper for the most common case.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

ekinanp commented on fuse/file.go:
```diff
@@ -3,103 +3,377 @@ package fuse
+	attr := plugin.Attributes(f.entry)
+	applyAttr(a, attr, defaultMode(f.entry))
+
+	if f.useLocalContent() || !f.isFileLikeEntry() {
```
Yeah, this makes more sense. Thanks.

ekinanp commented on docs/docs/index.md:
```diff
@@ -260,6 +261,34 @@ wash . ❯ grep "Hello" gcp/Wash/storage/some-wash-stuff/an\ example\ folder/sta
+
+Note that Wash distinguishes between file-like and non-file-like entries. An entry is file-like if it's readable and writable and defines its size; you can edit it like a file.
+
+If it doesn't define a size then it's non-file-like, and trying to open it with a ReadWrite handle will error; reads from it may not return data you previously wrote to it. You should check its documentation with the `docs` command for that entry's write semantics.
```
I'd mention the stuff about `vim` here, like "We also recommend not using editors with these entries to avoid weird behavior".

ekinanp commented on plugin/types.go:
```diff
@@ -15,7 +15,15 @@ or referenced via the API - and tools for controlling how its data is cached.
+interface allows sending data to the entry.
+
+Wash distinguishes between two different patterns for things you can read and write. It considers
+a "file-like" entry to be one with a defined size (so the `size` attribute is set when listing the
```
OK, I agree about using `size` alone for file-like things. But I think we should remove `size` from the `metadata/data.json` files because those aren't really "files" (editing them makes no sense) so they shouldn't be treated as such.

MikaelSmith replied:
I think editing metadata does make sense for some things: the metadata is a reflection of a mix of state and configuration, and you may be able to make changes to that configuration. Editing those values to change the configuration would make sense. This is especially true for everything in Kubernetes.

ekinanp replied:
Right, but what you write isn't going to strictly match what's read -- there isn't a 1:1 symmetry between the content because we output pretty-printed JSON. But I don't have strong feelings about keeping size around either so up to you.

MikaelSmith replied:
Ok, that makes sense.

ekinanp commented on plugin/types.go:
```diff
@@ -183,24 +191,33 @@ type Streamable interface {
+// entry is only Writable, then only full writes (starting from offset 0) are
+// allowed, anything else initiated by the filesystem will result in an error.
+//
+// It's up to the implementer to decide how much data integrity to guarantee.
```
What do you mean by "data integrity"?

MikaelSmith replied:
Primarily around whether writing volumes makes sure an fsync happens so that data is serialized to persistent storage before returning. Not sure how to summarize that.

ekinanp replied:
Ah ok. Sounds like an issue for file-like entries, which could be summed up as `plugin.Write(data); data == plugin.Read` (ensure that the next Read from the API will fetch the written data). Don't know if that implies "always call fsync" for volumes.

MikaelSmith replied:
I ended up removing this because I don't think any plugins should explicitly implement something like calling `sync` after writing. It represents a possible loss of data, but the machine would have to die right after and we'd likely get a connection error if that happened. Things like vim and databases will call Fsync, and we're currently lying to them in some cases that it did something, but for now I'd suggest don't run a database on Wash's filesystem.

ekinanp commented on plugin/types.go:
```diff
@@ -183,24 +191,33 @@ type Streamable interface {
+// Writable is an entry that we can write new data to. What that means can be
+// implementation-specific; it could be overwriting a file, submitting a
+// configuration change to an API, or writing data to a queue. It doesn't
+// support a concept of a partial write.
```
Should we add a reminder to plugin authors to include write-semantics for non-file-like entries in the description? Also holds for the external plugin docs.

MikaelSmith replied:
Done.

ekinanp commented on plugin/externalPluginEntry_test.go:
```diff
@@ -661,6 +661,109 @@ func (suite *ExternalPluginEntryTestSuite) TestListReadWithMethodResults() {
+
+var _ = internal.Command(&mockCommand{})
+
+func (suite *ExternalPluginEntryTestSuite) TestWrite() {
```
Why does Write have so many tests? Seems like it should be symmetric with other `InvokeAndWait` tests.

MikaelSmith replied:
I'll have to spend a little time on this tomorrow. I'm not sure the other tests are verifying all the assertions we expect. They also mock at a higher level (the script's InvokeAndWait call) than Write had available originally.

MikaelSmith committed:
Don't set size on non-file-like entries

Don't set size on console output and metadata files. The plugin system
knows how to get the size just as efficiently, and explicitly setting
the size implies they'll behave like files (which may not be true).

Signed-off-by: Michael Smith <michael.smith@puppet.com>
